Added script support and custom renderer

rion18 opened the pull request:
## Update

This PR now adds the ability to clean attachment files locally using the `cleanupLocalAttachmentFiles` option, adds SWC for faster jest tests, and updates all packages (where possible) to their latest versions.

## Description

- Added the ability of using Cosmere as an import inside of a script
- Added support for custom renderers
- Added multiple test cases
- Fixed some previous tests
- Updated changelog and readme

## Motivation and context

Using Cosmere is part of a pipeline I need to automate documentation creation using Confluence, which is why it's being used as a script instead of a CLI command.

However, some things remain to be rendered "easily", such as the ability to add width and height to an image. [marked](https://github.com/markedjs/marked) allows us to create any kind of renderer, which is why I thought of expanding this functionality into Cosmere.

## How has this been tested

I've created a suite of tests for the new `ObjectConfigLoader` as well as fixing some previous tests.

I've also installed it locally on one of the scripts using `npm install cosmere:@file://path_to_cosmere_fork` and I see that it works, as detailed in the readme.

rion18 committed:
Added script support and custom renderer

rion18 committed:
Ran linter

rion18 commented:
@mihaeu Hi! I really hope you can take a look at this PR soon. I love Cosmere, and I think adding more flexibility to its use can be a fantastic way of getting automated documentation. Expanding upon the original Markdown spec, as allowed by [marked](https://github.com/markedjs/marked) is also a great thing, that allows customizing tags in order to create even more beautiful docs. Thanks!

rion18 committed:
Local cleanup done

mihaeu commented:
@rion18 you are my absolute hero!!! I will try to look at this PR soon, promise. At first glance I love the changes :heart: 

Thanks so much for cleaning up along the way as well :muscle:

mihaeu commented:
Side-note: I was planning on adding image sizes as well, but I still think adding scripting support would be very cool.

rion18 commented:
Hi! Glad you're up to checking this PR. Sadly image sizes are ONLY part of [kramdown](https://kramdown.gettalong.org/syntax.html#images) which is not supported by [marked](https://github.com/markedjs/marked). 

However, this can be easily achieved by the custom renderer trivially. Here's the implementation I'm using:

```js
const cosmere = require('cosmere/dist/src/lib');
const ConfluenceRenderer = require('cosmere/dist/src/ConfluenceRenderer').default;

const IMAGE_URL_REGEX = /^(.+) \| (.+)$/;
const WIDTH_REGEX = /(w|width)=(\d+)px/;
const HEIGHT_REGEX = /(h|height)=(\d+)px/;

class CustomRenderer extends ConfluenceRenderer {

  constructor(options, config, page) {
    super(options, config, page);
  }

  imageTag(href, width, height, title, text) {
    if (href.startsWith('http')) {
      return `<ac:image ac:width="${width || ''}" ac:height="${height || ''}" ac:title="${title || ''}" ac:alt="${text || ''}"><ri:url ri:value="${href}"/></ac:image>`;
    }
    return `<ac:image ac:width="${width || ''}" ac:height="${height || ''}" ac:title="${title || ''}" ac:alt="${text || ''}"><ri:attachment ri:filename="${href}"/></ac:image>`;
  }

  image(href, title, text) {
    const match = IMAGE_URL_REGEX.exec(href);
    if (!match) {
      return this.imageTag(href, '', '', title, text);
    }
    const heightWidth = match[2].replace(' ', '');
    const widthMatch = WIDTH_REGEX.exec(heightWidth);
    const width = !widthMatch ? '' : widthMatch[2];
    const heightMatch = HEIGHT_REGEX.exec(heightWidth);
    const height = !heightMatch ? '' : heightMatch[2];
    return this.imageTag(match[1], width, height, title, text);
  }
}
```

And in my markdown I use:

```md
![image](../myAwesomeImage.svg | h=420px)
![image](../myAwesomeImage.svg | w=640px)
![image](../myAwesomeImage.svg | h=420px w=512px) // introduces a cropped image, but setting one of the other fixes this

![image](../myAwesomeImage.svg | height=420px)
![image](../myAwesomeImage.svg | width=640px)
![image](../myAwesomeImage.svg | height=420px width=512px) // introduces a cropped image, but setting one of the other fixes this
```

The possibilities are endless.

rion18 commented:
Also... we should update Axios since running npm install warns of a security vulnerability.

mihaeu commented:
Hey there, sorry for taking so long. Last week was crazy.

I love everything, but there's only one thing I'm not sure about. Why are you clearing the pictures from the cache every time?

rion18 commented:
No worries. 

I actually thought the pictures were a bug, since they are always stored in the root of project (instead of the cache folder, at least on my project), generating unwanted artifacts. However, you don't need those in there, since you do a check against the attachment's size in the Confluence doc. 

Proof of this is, if you use the version on my PR with multiple attachments, and then update the document (without updating the attachments), they do NOT get updated on Confluence, which is expected. If you update a single attachment, then ONLY that attachment gets updated.

rion18 committed:
Added cleanupLocalAttachmentFiles option

rion18 committed:
Added MainCommand tests that checks internal parameters

rion18 commented:
Good news @mihaeu.
I've added the ability to enable/disable caching local attachment files using the `cleanupLocalAttachmentFiles` option, on both `FileConfig` and `ObjectConfig`. 
I've also added a test for the MainCommand CLI, which tests the paramaters that are passed down to `updatePage`.

rion18 committed:
Updated all possible packages

rion18 committed:
Added swc as a transform function for jest

rion18 commented:
Some other news. I've updated ALL of the dependencies (except inquire which is now an ESM module), and I've added SWC as a transform function for jest. Tests are now BLAZING fast.

rion18 commented:
Any news about the PR? Is there anything else I need to update/change? @mihaeu ?

mihaeu commented:
Hey @rion18, thanks for all the stuff you added. There's only one question and one remark:

* Is adding swc worth having the extra dependency? How much speed are you gaining there?

* I actually liked the point you were making regarding the cached files. I would go one step further and just remove them immediately after uploading them, since they just bloat up the cache (which is especially annoying on CI runners). What do you think?

I could take this over, since you've already done so much, and I want to make sure your work is merged soon. Sounds good to you?

rion18 commented:
Well since it's a dev dependency it shouldn't bear as much impact as having a bundled one. We would need to benchmark the speed increase in a CI environment. 

In my setup, swc tests take around 1.4 sec versus 2.5 with ts-jest, but it's a veeeeery relative way to "measure" this. I don't think it's a dealbreaker to use it or not since there are not many tests.

As for the file cleanup, I agree, we can totally remove the files after upload and potentially even get rid of the "cleanupLocalAttachmentFiles" option, if you'd like.

mihaeu committed:
Remove temporary attachments after upload

Fix build after axios upgrade

mihaeu commented:
I noticed that the build is no longer breaking since the upgrade of the dependencies. I don't want to downgrade however. Axios is working again, but the axios-file dependency is still broken. I don't like axios on node, so I think I'll just get rid of it and then we can go and merge this.

mihaeu committed:
Remove temporary attachments after upload

Fix build after axios upgrade

mihaeu commented:
I got rid of `axios-file` and realized that `--force` didn't work for attachments, so I fixed that as well.

Wanna have a last look? Are you okay with me listing you in the `README.md` as a contributor @rion18 ?

mihaeu commented:
I realized I have a lot of work to do over the holidays because test coverage is horrible, ESM/commonjs setup sucks etc.

rion18 commented:
Sure thing, I'd be glad to be marked as a contributor. I can help out with tests and stuff like that as well if you'd like.

rion18 committed:
Removed resolutions section in package json

rion18 commented:
Just looked at the commits and they look good. I added one removing "resolutions" under package.json

rion18 commented:
Awesome! @mihaeu Thanks for merging this into main. Don't forget to create the new release ;)
